### ISSUE TITLE
instrumentExecutionInput not updating ExecutionInput reference

### ISSUE THREAD

**Issue by bkuberek** [graphql-java/graphql-java#1667]
**Describe the bug**

See the following lines in GraphQL.java, 
https://github.com/graphql-java/graphql-java/blob/9e833202b8dec147b37ba128ff3e3e3b37df26e3/src/main/java/graphql/GraphQL.java#L492-L497

Notice that the `InstrumentationState` is created with an `ExecutionInput` with certain properties.

Then the `executionInput` is overriden with a new instance but the state still holds the old reference.

Anywhere in the stack where you ask the `InstrumentationState` for the `ExecutionInput` it will return the original one instead of the new one.

This makes it impossible to transform `ExecutionInput` in `Instrumentation.instrumentExecutionInput` without also overriding `Instrumentation.createState` to set the state reference as well.

**To Reproduce**

```java
graphQL = graphQL.instrumentation(new LexikonInstrumentation())

ExecutionInput executionInput = ExecutionInput.newExecutionInput()
    .query(query)
    .operationName(operationName)
    .context(exchange)
    .root(exchange)
    .variables(variables)
    .build();
ExecutionResult executionResult = graphQL.execute(executionInput);
```

The Instrumentation class

```java
public static class LexikonInstrumentation extends DataLoaderDispatcherInstrumentation {
  @Override
  public InstrumentationState createState(final InstrumentationCreateStateParameters parameters) {
    // parameters.getExecutionInput() is the original ExecutionInput
    // and it has an empty DataLoaderRegisatry
    return new DataLoaderDispatcherInstrumentationState(LOG,
        transformExecutionInput(parameters.getExecutionInput()).getDataLoaderRegistry());
  }

  @Override
  public ExecutionInput instrumentExecutionInput(final ExecutionInput executionInput,
                                                 final InstrumentationExecutionParameters parameters) {
    DataLoaderRegistry registry = new DataLoaderRegistry();

    registry.register(
        "WorkplacePost",
        DataLoader.newMappedDataLoader(workplacePostLoader
    );

    return executionInput.transform(builder -> builder.dataLoaderRegistry(registry));
  }
}
```

**Comment by bbakerman** [comment]
This is a tricky one to "fix" because I think it is working as designed.

The `Instrumentation` interface is a life cycle interface.  It starts at the begging of the the request and the first thing it does it allow the `Instrumentation` implementation to create state for itself (with the starting ExecutionInput as you noted).  One thing an implementation can choose to do it "enhance" the ExecutionInput say.  If later stages need this new ExecutionInput shape reflected in the state then it MUST do that itself.

I notice in your examppe you use a custom `Instrumentation` implementation that puts new values into the ExecutionInput.  

`DataLoaderDispatcherInstrumentation` has not be designed to be extended.  I suggest you build up this code

```
DataLoaderRegistry registry = new DataLoaderRegistry();

    registry.register(
        "WorkplacePost",
        DataLoader.newMappedDataLoader(workplacePostLoader
    );
```

Outside the when you originally build the execution input the first time

**Comment by bkuberek** [comment]
The only reason I am doing it this way is because I am using a framework that currently doesn't permit adding registry to execution input before calling execute. We might fix that instead.

**Comment by bbakerman** [comment]
I looked more into this and I think it is fair that IF an Instrumentation decides to enhance the ExecutionInput (and hence DataLoaderRegistry instance) then it should be picked up

I have created a PR that addresses this

**Comment by andimarek** [comment]
fixed by #1668
